src/sagas/task.saga.js

```diff
@@ -0,0 +1,339 @@
+            yield put(
+              receiveDataFromTaskExecution(task, message.text, message.isError)
+            );
+            break;
+          case 'stderr':
+            yield put(receiveDataFromTaskExecution(task, message.text));
+            break;
+          case 'exit':
+            yield call(displayTaskComplete, task);
+            yield put(
+              completeTask(task, message.timestamp, message.wasSuccessful)
+            );
+            break;
+        }
+      }
+    } finally {
+      /* child process has completed */
+    }
+  } catch (err) {
+    // TODO: Error handling (this can happen if the first 15 ports are occupied,
+    // or if there's some generic Node error
+    console.error(err);
+  }
+}
+
+export function* taskRun({ task }: { task: Task }): Saga<void> {
+  const project = yield select(getProjectById, task.projectId);
+  const projectPath = yield select(getPathForProjectId, task.projectId);
+  const { name } = task;
+
+  // TEMPORARY HACK
+  // By default, create-react-app runs tests in interactive watch mode.
+  // This is a brilliant way to do it, but it's interactive, which won't
+  // work as-is.
+  // In the future, I expect "Tests" to get its own module on the project
+  // page, in which case we can support the interactive mode, except with
+  // descriptive buttons instead of cryptic letters!
+  // Alas, this would be mucho work, and this is an MVP. So for now, I'm
+  // disabling watch mode, and doing "just run all the tests once" mode.
+  // This is bad, and I feel bad, but it's a corner that needs to be cut,
+  // for now.
+  const additionalArgs = [];
+  if (project.type === 'create-react-app' && name === 'test') {
+    additionalArgs.push('--coverage');
+  }
+
+  const child = yield call(
+    [childProcess, childProcess.spawn],
+    PACKAGE_MANAGER_CMD,
+    ['run', name, ...additionalArgs],
+    {
+      cwd: projectPath,
+      env: getBaseProjectEnvironment(projectPath),
+    }
+  );
+
+  // TODO: Does the renderer process still need to know about the child
+  // processId?
+  yield put(attachTaskMetadata(task, child.pid));
+
+  // When this application exits, we want to kill this process.
+  // Send it up to the main process.
+  yield call([ipcRenderer, ipcRenderer.send], 'addProcessId', child.pid);
+
+  const stdioChannel = createStdioChannel(child, {
+    stdout: emitter => data => {
+      // The 'eject' task prompts the user, to ask if they're sure.
+      // We can bypass this prompt, as our UI already has an alert that
+      // confirms this action.
+      // TODO: Eject deserves its own Redux action, to avoid cluttering up
+      // this generic "RUN_TASK" action.
+      // TODO: Is there a way to "future-proof" this, in case the CRA
+      // confirmation copy changes?
+      const isEjectPrompt = data
+        .toString()
+        .includes('Are you sure you want to eject? This action is permanent');
+
+      if (isEjectPrompt) {
+        sendCommandToProcess(child, 'y');
```

Comment:
so glad we can do this now :D I know it's unrelated to this change, it just makes me really happy, for the future test possibilities